compiler/rustc_smir/src/rustc_smir/mod.rs

```diff
@@ -79,6 +80,48 @@ impl<'tcx> Tables<'tcx> {
         };
         !must_override && self.tcx.is_mir_available(def_id)
     }
+
+    fn to_fn_def(&mut self, tcx: TyCtxt<'tcx>, def_id: DefId) -> Option<FnDef> {
+        if matches!(tcx.def_kind(def_id), DefKind::Fn | DefKind::AssocFn) {
+            Some(self.fn_def(def_id))
+        } else {
+            None
+        }
+    }
+
+    fn to_static(&mut self, tcx: TyCtxt<'tcx>, def_id: DefId) -> Option<StaticDef> {
+        matches!(tcx.def_kind(def_id), DefKind::Static { .. }).then(|| self.static_def(def_id))
+    }
+
+    fn filter_map_items<T, F>(&mut self, tcx: TyCtxt<'tcx>, krate: CrateNum, func: F) -> Vec<T>
+    where
+        F: Fn(&mut Tables<'tcx>, TyCtxt<'tcx>, DefId) -> Option<T>,
+    {
+        let crate_def = krate.as_def_id();
+        let mut queue = vec![crate_def];
+        queue.extend_from_slice(tcx.trait_impls_in_crate(krate));
+        let mut visited = HashSet::default();
+        let mut result = vec![];
+        while let Some(def_id) = queue.pop() {
```

Comment:
Frankly, actually, now that I think of it, I wonder if this whole approach is somewhat roundabout, and instead the right way of doing this would simply be to simply loop over all of the `DefIndex`es of the given crate.

That's basically what [`iter_local_def_id`](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_middle/ty/context.rs.html#1865-1888) does. Perhaps we should just expose some sort of `num_foreign_definitions` that takes a `CrateNum` and maps it to the number of definitions... 

maybe you can extract that from the [`CrateMetadata`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_metadata/rmeta/decoder/struct.CrateMetadata.html), or maybe it needs to be encoded...

Reply:
That would be much easier. Let me give it a try. Thanks